src/library/commissioner_impl.cpp

```diff
@@ -1955,6 +1955,288 @@ ByteArray CommissionerImpl::GetCommissionerDatasetTlvs(uint16_t aDatasetFlags)
+    {
+        const ByteArray &value = childIpv6Address->GetValue();
+        SuccessOrExit(error = DecodeChildIpv6AddressList(dataset.mChildIpv6AddrsInfoList, value));
+        dataset.mPresentFlags |= NetDiagData::kChildIpv6AddressBit;
+    }
+
+    aNetDiagData = dataset;
+
+exit:
+    return error;
+}
+
+Error CommissionerImpl::DecodeIpv6AddressList(std::vector<std::string> &aAddrs, const ByteArray &aBuf)
+{
+    Error  error;
+    size_t length = aBuf.size();
+    size_t offset = 0;
+    while (offset < length)
+    {
+        VerifyOrExit(offset + kIpv6AddressBytes <= length, error = ERROR_BAD_FORMAT("premature end of IPv6 Address"));
+        Address          addr;
+        const ByteArray &value = {aBuf.data() + offset, aBuf.data() + offset + kIpv6AddressBytes};
+        SuccessOrExit(error = addr.Set(value));
+        aAddrs.emplace_back(addr.ToString());
+        offset += kIpv6AddressBytes;
+    }
+exit:
+    return error;
+}
+
+Error CommissionerImpl::DecodeChildIpv6AddressList(std::vector<ChildIpv6AddressInfo> &aChildIpv6AddressInfoList,
+                                                   const ByteArray                   &aBuf)
+{
+    Error                error;
+    size_t               length = aBuf.size();
+    ChildIpv6AddressInfo childIpv6AddrsInfo;
+
+    VerifyOrExit((length - kRloc16Bytes) % kIpv6AddressBytes == 0,
+                 error = ERROR_BAD_FORMAT("premature end of Child IPv6 Address"));
+    childIpv6AddrsInfo.mRloc16 = utils::Decode<uint16_t>(aBuf.data(), kRloc16Bytes);
+
+    SuccessOrExit(error = DecodeIpv6AddressList(childIpv6AddrsInfo.mAddrs, {aBuf.begin() + kRloc16Bytes, aBuf.end()}));
+    aChildIpv6AddressInfoList.emplace_back(childIpv6AddrsInfo);
+exit:
+    return error;
+}
+
+Error CommissionerImpl::DecodeModeData(ModeData &aModeData, const ByteArray &aBuf)
+{
+    Error  error;
+    size_t length = aBuf.size();
+    VerifyOrExit(length == 1, error = ERROR_BAD_FORMAT("incorrect size of ModeData"));
+    aModeData.mIsRxOnWhenIdleMode          = (aBuf[0] & 0x01) != 0;
+    aModeData.mIsMtd                       = (aBuf[0] & 0x02) != 0;
+    aModeData.mIsStableNetworkDataRequired = (aBuf[0] & 0x04) != 0;
+exit:
+    return error;
+}
+
+Error CommissionerImpl::DecodeChildTable(std::vector<ChildTableEntry> &aChildTable, const ByteArray &aBuf)
+{
+    Error  error;
+    size_t length = aBuf.size();
+    size_t offset = 0;
+
+    while (offset < length)
+    {
+        ChildTableEntry entry;
+        VerifyOrExit(offset + kChildTableEntryBytes <= length,
+                     error = ERROR_BAD_FORMAT("premature end of Child Table"));
+        entry.mTimeout             = 1 << (((aBuf[offset] & 0xF8) >> 3) - 4);
+        entry.mIncomingLinkQuality = (aBuf[offset] & 0x06) >> 1;
+        entry.mChildId             = ((aBuf[offset] & 0x01) << 9) | aBuf[offset + 1];
+        SuccessOrExit(error = DecodeModeData(entry.mModeData, {aBuf.begin() + offset + 2, aBuf.begin() + offset + 3}));
```

Comment:
Can you document the TLV value format here? It's hard to follow the code.

Reply:
done